PUN PUN DRIP

13spacemen opened the pull request:
![punpunclothes](https://user-images.githubusercontent.com/46101244/179372695-b24180d9-3fb0-4e82-a676-c950471e8426.png)

Pun Pun spawns with a waiter suit on
code-wise adds support for monkeys having their own clothes sprite variations like how digitigrade limbs do

Closes https://github.com/tgstation/tgstation/issues/66465

:cl:
imageadd: Pun Pun wears drippy clothes now. Swag.
fix: Becoming a digitigrade species should properly drop your shoes if they don't fit
/:cl:

13spacemen committed:
adds var/bodytypes_whitelist to make clothes have a bodytype whitelist
adds monkey specific clothing files
adds clothes for punpun which he spawns with

Fikou commented on code/modules/mob/living/carbon/human/monkey/monkey.dm:
```diff
@@ -34,6 +34,18 @@
+	can_adjust = FALSE
+	female_sprite_flags = NO_FEMALE_UNIFORM
+	species_exception = list(/datum/species/monkey)
+	bodytypes_restricted = BODYTYPE_MONKEY
```
restricted sounds like youre allowing everyone except monkeys to wear it. why not just whitelist

Kapu1178 commented on code/modules/mob/living/carbon/human/species.dm:
```diff
@@ -805,6 +811,10 @@ GLOBAL_LIST_EMPTY(features_by_species)
 	if(slot in no_equip)
 		if(!I.species_exception || !is_type_in_list(src, I.species_exception))
 			return FALSE
+	if(!isnull(I.bodytypes_whitelist))
+		if(!(I.bodytypes_whitelist & H.dna.species.bodytype))
+			to_chat(H, span_warning("Your species cannot wear this item!"))
+			return FALSE
 
```
this is dumb. why not just make the definition ALL?

Kapu1178 commented on code/modules/mob/living/carbon/human/species.dm:
```diff
@@ -432,6 +432,12 @@ GLOBAL_LIST_EMPTY(features_by_species)
 		var/obj/item/thing = C.get_item_by_slot(slot_id)
 		if(thing && (!thing.species_exception || !is_type_in_list(src,thing.species_exception)))
 			C.dropItemToGround(thing)
+	if(ishuman(C))
```
pointless check. The below proc is carbon-level and will automatically call the human-level version when invoked.

13spacemen committed:
minor code edits, reduces amount of checks

Mothblocks commented on code/modules/mob/living/carbon/human/species.dm:
```diff
@@ -432,6 +432,10 @@ GLOBAL_LIST_EMPTY(features_by_species)
 		if(thing && (!thing.species_exception || !is_type_in_list(src,thing.species_exception)))
 			C.dropItemToGround(thing)
+	if(C.dna)
+		for(var/obj/item/the_item in C.get_equipped_items(include_pockets = TRUE))
+			if(!(the_item.bodytypes_whitelist & C.dna.species.bodytype))
+				C.dropItemToGround(the_item)
```
```suggestion
		for(var/obj/item/equipped_item in C.get_equipped_items(include_pockets = TRUE))
			if(!(equipped_item.bodytypes_whitelist & C.dna.species.bodytype))
				C.dropItemToGround(equipped_item)
```

Mothblocks committed:
Use clearer name

Mothblocks commented on code/modules/mob/living/carbon/human/species.dm:
```diff
@@ -805,6 +809,9 @@ GLOBAL_LIST_EMPTY(features_by_species)
 		if(!I.species_exception || !is_type_in_list(src, I.species_exception))
 			return FALSE
+	if(!(I.bodytypes_whitelist & H.dna.species.bodytype))
+		to_chat(H, span_warning("Your species cannot wear this item!"))
```
I'm noticing its also in this proc, but something named `can_*` should never have side effects.

I'll let this go because the other ones have it and a huge refactor would be non-trivial if you can check something for me--if I, a monkey, have an item I can't equip in my hand, and hover it over its slot (to see the red/green silhouette), does it spit a message in the chat?

13spacemen replied:
removed var/bodytypes_whitelist

JohnFulpWillard commented on code/modules/mob/living/carbon/human/monkey/monkey.dm:
```diff
@@ -34,6 +34,18 @@
+	can_adjust = FALSE
+	female_sprite_flags = NO_FEMALE_UNIFORM
+	species_exception = list(/datum/species/monkey)
+	bodytypes_whitelist = BODYTYPE_MONKEY
```
Do we need to use this at all? Digitigrade clothings (and now monkey clothings) change their icon already when they are equipped to a person with that bodytype. Couldn't this just be added to the waiter's suit as the monkey variation of it??

13spacemen replied:
Made it use the human waiter's suit instead

Twaticus replied:
does that mean we can remove this 
![image](https://user-images.githubusercontent.com/46540570/179638838-b5a0d60c-cd9c-4382-b7e1-94177cba349e.png)

13spacemen replied:
@Twaticus good catch will delete

13spacemen committed:
removed punpun monkey-specific suit, let monkeys wear the regular waiter suit

JohnFulpWillard commented on code/game/objects/items.dm:
```diff
@@ -148,6 +148,8 @@ GLOBAL_DATUM_INIT(welding_sparks, /mutable_appearance, mutable_appearance('icons
 	var/list/attack_verb_simple
 	///list() of species types, if a species cannot put items in a certain slot, but species type is in list, it will be able to wear that item
 	var/list/species_exception = null
+	///A bitfield of bodytypes that the item can be worn by only.
+	var/bodytypes_whitelist = ALL
```
this isn't needed anymore with the new changes

JohnFulpWillard commented on code/modules/mob/living/carbon/human/monkey/monkey.dm:
```diff
@@ -76,6 +76,7 @@ GLOBAL_DATUM(the_one_and_only_punpun, /mob/living/carbon/human/species/monkey/pu
 		equip_to_slot_or_del(new relic_hat, ITEM_SLOT_HEAD)
 	if(relic_mask)
 		equip_to_slot_or_del(new relic_mask, ITEM_SLOT_MASK)
+	equip_to_slot_or_del(new /obj/item/clothing/under/suit/waiter(src), ITEM_SLOT_ICLOTHING)
```
Do we want pun pun to start with this or have bartenders buy the suit from their vendor?

13spacemen replied:
starting with it is better, you can remove it from Pun Pun if you don't like it

JohnFulpWillard commented on code/modules/mob/living/carbon/human/species.dm:
```diff
@@ -432,6 +432,10 @@ GLOBAL_LIST_EMPTY(features_by_species)
 		var/obj/item/thing = C.get_item_by_slot(slot_id)
 		if(thing && (!thing.species_exception || !is_type_in_list(src,thing.species_exception)))
 			C.dropItemToGround(thing)
+	if(C.dna)
+		for(var/obj/item/equipped_item in C.get_equipped_items(include_pockets = TRUE))
+			if(!(equipped_item.bodytypes_whitelist & C.dna.species.bodytype))
+				C.dropItemToGround(equipped_item)
```
Same as first review, bodytyoes_whitelist isn't used anymore so we don't need to add all this.

13spacemen committed:
remove unused var/bodytypes_whitelist and redundant punpun clothing obj icon

JohnFulpWillard reviewed (commented):
Do we want monkeys being able to wear clothes be a normal thing, or this suit be a sole exception?

If we want it to be normal, we can allow monkeys to wear jumpsuits as a species, but limit it (like digis) to those that have ``CLOTHING_MONKEY_VARIATION``, if it is something you would prefer (again, if it's what you want).

13spacemen commented:
@JohnFulpWillard It should be a normal thing but code-wise isn't it the same as var/species_exception? We'll be deleting that on the monkey clothes but we'll still have to add a check to human/species equipping code

13spacemen committed:
removed the species_exception on waiter suit
uniforms on monkeys are no longer no_equip
added checks for monkey bodypart and clothing match

13spacemen committed:
reduced code duplication in on_species_gain()
added disable_warning to ICLOTHING monkey check

13spacemen commented on code/modules/mob/living/carbon/human/species.dm:
```diff
@@ -441,6 +441,12 @@ GLOBAL_LIST_EMPTY(features_by_species)
+	var/obj/item/uniform = C.get_item_by_slot(ITEM_SLOT_ICLOTHING)
+	if(chest && (chest.bodytype & BODYTYPE_MONKEY) && uniform)
+		if(!(uniform.supports_variations_flags & CLOTHING_MONKEY_VARIATION))
+			C.dropItemToGround(uniform)
```
done

13spacemen commented:
@JohnFulpWillard not sure why it fails linters, maps compile fine, any idea?

JohnFulpWillard commented on code/modules/mob/living/carbon/human/species.dm:
```diff
@@ -428,19 +428,19 @@ GLOBAL_LIST_EMPTY(features_by_species)
 	replace_body(C, src)
 	regenerate_organs(C, old_species, visual_only = C.visual_only_organs)
 
+	for(var/obj/item/equipped_item in C.get_equipped_items(include_pockets = TRUE))
+		var/equipped_item_slot = C.get_slot_by_item(equipped_item)
+		if(!can_equip(equipped_item, equipped_item_slot, H = C,  bypass_equip_delay_self = TRUE))
+			C.dropItemToGround(equipped_item)
```
There's a few problems here:

1- ``get_equipped_items`` gets every single item on a mob, but ``get_slot_by_item`` doesn't take pockets/suit storage into account, so they end up being null, we don't actually have to check pockets because we already check the jumpsuit/exosuit, and if those come off, their items go off with them.
2- can_equip returns FALSE if you already have a piece of clothing in that slot, and since we already have a full set of clothes, it causes us to drop everything.

Lastly, clothes have ``slot_flags`` which you can take advantage of here, bypassing the need for ``get_slot_by_item``.

13spacemen replied:
adjusted this

13spacemen committed:
added var to can_equip() to bypass equipped clothing check
minor code edits

13spacemen committed:
INVOKE_ASYNC
minor code edits
